### PR TITLE
Add go-version-from-file option

### PR DESCRIPTION
**Pull request by jojo43**
I added an option `go-version-from-file` as suggested in https://github.com/actions/setup-go/issues/23#issuecomment-606215099 .

### CODE REVIEW

**Comment by radeksimko** [comment]
👋🏻  Is there anything in particular blocking this PR from being merged?

**Comment by bflad** [comment]
Echoing the above, is there anything the community can do to help get this enhancement reviewed? It would be very useful to replace a workaround we use in a few dozen repositories that my team manages.

**Comment by driktea** [comment]
goodluk so I like to aplikasion

**Review by viccastro0916 — APPROVED** [review]

**Review by viccastro0916 — APPROVED** [review]

**Comment by Weerasak2020th** [comment]
> ฉันจะเพิ่มตัวเลือก`go-version-from-file`ที่แนะนำใน[# 23 ](https://github.com/actions/setup-go/issues/23#issuecomment-606215099)(ความคิดเห็น)

**Comment by leb4r** [comment]
> @ccremer I agree that such thing would be helpful, but it seems to me like a separate feature, as the complexity of pulling any content out of `go.mod` (reliably) requires parsing it.

I second this, some repositories do not implement a `go.mod` file or have no need for one. Any particular reason why this hasn't been merged yet?

**Comment by sifanjoraboris** [comment]
Wow Best

**Comment by tisba** [comment]
The Go version is already included in `go.mod`, so a dedicated Go version file is only needed if you really need to have an exact patch release (which sometimes can be useful of corse). In most cases I'd argue that the best practise should be to read `go.mod` and use the latest version.

Essentially that would be something like this: `go mod edit -json | jq -r .Go`.

This could be an alternative option as well. Instead "from file" a "from go mod" or something 🤔 

See https://github.com/actions/setup-go/issues/174#issuecomment-1096139190

**Comment by radeksimko** [comment]
> The Go version is already included in go.mod

I think those are two different things. The version in `go.mod` is a compatibility _constraint_ whereas `setup-go` cares about what exact version to use, which presumably should be either equal or higher than the one in `go.mod`. In many cases though compatibility constraint is not how you'll pick versions to use in CI AFAICT. Typically folks develop and test against particular Go versions, not just "any latest" which matches compatibility constraint.

**Comment by IvanZosimov** [comment]
Hi, @jojo43 👋 We are planning to implement this feature in the next releases of the action, but your PR is a little bit outdated and may be updated. Are you still interested in contributing 🚀 ?

**Comment by jojo43** [comment]
@IvanZosimov 
Yes! :smiley: 
What changes are needed?

**Review by IvanZosimov — CHANGES_REQUESTED** [review]
Hi @jojo43 👋 Thank you very much for your update, it looks great, just a few changes are needed and it will brilliant ! Could I also ask you to sync your branch with the upstream branch and reattach this PR to the `main` branch (now it's attached to the master)?

**Comment by jojo43** [comment]
@IvanZosimov 
Please take a look again.

File: action.yml [hunk 1]
```diff
@@ -4,6 +4,8 @@ author: 'GitHub'
 inputs: 
   go-version:
     description: 'The Go version to download (if necessary) and use. Supports semver spec and ranges.'
+  go-version-from-file:
+    description: Path to the file with the Go version.  go-version overwrites this.
```
[the remark this answers is not in the document]
Comment (radeksimko) [reply action.yml]: @IvanZosimov FWIW as an end user I would find the original description more accurate. While I guess none of us can predict what % of users will end up specifying `go.mod` and what other % `.go-version`, but I assume (hope) that the intention is to support both and so the description should reflect that?

**Comment by dmitry-shibanov** [comment]
Hello @jojo43. Could you please add new job in `versions.yml` with go-version-file input.

**Comment by panchoh** [comment]
> > The Go version is already included in go.mod
> 
> I think those are two different things. The version in `go.mod` is a compatibility _constraint_ whereas `setup-go` cares about what exact version to use, which presumably should be either equal or higher than the one in `go.mod`. In many cases though compatibility constraint is not how you'll pick versions to use in CI AFAICT. Typically folks develop and test against particular Go versions, not just "any latest" which matches compatibility constraint.

Thanks a bunch to @jojo43 for pushing this!

Nevertheless, please take into consideration @radeksimko remarks, because they are spot-on, IMHO.  For a CI-CD workflow, you typically want to select the exact point revision you are using, not what happens to be the last one available (if you want reproducible builds, that is).  Otherwise you can end up with a non-functioning build, and be flabbergasted about why.  For instance, in the 1.16.x or 1.17.x series there was a point release that changed the behavior of the web server, requiring changes to code using it, otherwise your program it wouldn't build.

I wholeheartedly recommend the approach shown [here](https://github.com/actions/setup-go/issues/23#issuecomment-647497896), where the Go version is read from a `Dockerfile`.  It's genius! Because then, you can trivially configure **dependabot** to look for updates to the `Dockerfile`, and since the Go project publishes updated Docker images as part of each release, you can keep your build updated with very little effort, and in a controlled fashion. Dependabot will produce a PR with the updated Dockerfile pointing to the latest Go release available. You can accept (or not) the PR, and get yourself an updated build. If you have tests, the PR will run those, and you get feedback on the effects of the change.  Works like a charm!

My 2¢.

Thanks again!

**Review by brcrista — APPROVED** [review]

**Review by marko-zivic-93 — APPROVED** [review]

**Comment by IvanZosimov** [comment]
> @IvanZosimov Please take a look again.

Hi, @jojo43 👋 Thanks a lot for your updates! Could you also update documentation (the README.md file) and add there this chapter:

```
## Getting go version from the go.mod file

The `go-version-file` input accepts a path to a `go.mod` file containing the version of Go to be used by a project. As the `go.mod` file contains only major and minor (e.g. 1.18) tags, the action will search for the latest available patch version sequentially in the runner's directory with the cached tools, in the [version-manifest.json](https://github.com/actions/go-versions/blob/main/versions-manifest.json) file or at the go servers. 

If both the `go-version` and the `go-version-file` inputs are provided then the `go-version` input is used.
> The action will search for the `go.mod` file relative to the repository root

```yaml
steps:
- uses: actions/checkout@v3
- uses: actions/setup-go@v3
  with:
    go-version-file: '**/go.mod'
- run: go version
```

File: action.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (IvanZosimov) [reply action.yml]: Hi, @radeksimko 👋 ! Thank you, that's definitely a good point! We are going to update setup-go documentation with the new chapter related to this functionality with a more broad description of the input.

**Review by dmitry-shibanov — APPROVED** [review]

**Review by marko-zivic-93 — APPROVED** [review]

**Review by vsafonkin — APPROVED** [review]

**Comment by IvanZosimov** [comment]
Hi, @minamijoyo 👋 Thanks a lot for your concerns, actually you are right there, the scope of this PR is to add retrieving Go version from `go.mod` file functionality. Nevertheless, you can specify path to `.go-version` file in the input `go-version-file` and our action will read the version from it. We will update the documentation with a note about that later after merging this PR.
 
Cheers!

**Review by IvanZosimov — APPROVED** [review]

**Review by marko-zivic-93 — APPROVED** [review]

**Pull request merged**